src/Utilities/struct_of_constraints.jl

```diff
@@ -365,13 +365,21 @@ function struct_of_constraint_code(struct_name, types, field_types = nothing)
             ),
         )
     end
+    # Here we make an internal constructor for our type. It needs to be an
+    # internal constructor to avoid someone trying to provide
+    # `typed_struct(args...)` which will throw an error about `T not defined`.
     constructor_code = :(function $typed_struct() where {$T}
-        return $typed_struct(0, $([:(nothing) for _ in field_types]...))
+        return new{$T}(0, $([:(nothing) for _ in field_types]...))
     end)
     if type_parametrized
+        # A bit confusing: we need to add the extra type parameters to `{T}`.
+        # Here's the `where {T}`
         append!(constructor_code.args[1].args, field_types)
+        # And here's the `new{T}`
+        append!(constructor_code.args[2].args[end].args[1].args[1].args, field_types)
     end
-    push!(code.args, constructor_code)
+    # Here's we we put the inner constructor.
+    push!(code.args[2].args[3].args, constructor_code)
```

Comment:
Okay. Previously, this code made an external constructor. This could lead to errors like
```Julia
julia> module TestFoo

       struct Foo{T}
           x::Union{Nothing,Vector{T}}
       end

       function Foo{T}() where {T}
           return Foo{T}(nothing)
       end

       end
Main.TestFoo

julia> TestFoo.Foo(nothing)
ERROR: UndefVarError: T not defined
Stacktrace:
 [1] Main.TestFoo.Foo(x::Nothing)
   @ Main.TestFoo ./REPL[9]:4
 [2] top-level scope
   @ REPL[10]:1

julia> TestFoo.Foo()
ERROR: MethodError: no method matching Main.TestFoo.Foo()
Closest candidates are:
  Main.TestFoo.Foo(::Union{Nothing, Vector{T}}) where T at REPL[9]:4
Stacktrace:
 [1] top-level scope
   @ REPL[11]:1

julia> TestFoo.Foo{Float64}()
Main.TestFoo.Foo{Float64}(nothing)
```

This new code creates an internal constructor:
```Julia
julia> module TestFoo

       struct Foo{T}
           x::Union{Nothing,Vector{T}}
           function Foo{T}() where {T}
               return new{T}(nothing)
           end
       end

       end
Main.TestFoo

julia> TestFoo.Foo(nothing)
ERROR: MethodError: no method matching Main.TestFoo.Foo(::Nothing)
Stacktrace:
 [1] top-level scope
   @ REPL[6]:1

julia> TestFoo.Foo()
ERROR: MethodError: no method matching Main.TestFoo.Foo()
Stacktrace:
 [1] top-level scope
   @ REPL[7]:1

julia> TestFoo.Foo{Float64}()
Main.TestFoo.Foo{Float64}(nothing)
```

Reply:
#1735